hashicorp/dev-portal
# TFE Releases - versioning bug fixes

## 🔗 Relevant links



- [Preview link](https://dev-portal-git-heat-fixtfe-dropdown-hashicorp.vercel.app/terraform/enterprise/releases) 🔎
- [Asana task](https://app.asana.com/0/1206176289707208/1206567316230774/f) 🎟️

## 🗒️ What

- Updated the following functions to account for the different versioning style in TFE Release docs, which may have two versions or one (e.g. `'/terraform/enterprise/v202208-2/releases/2022/v202208-2'` and `'/terraform/enterprise/releases/2022/v202208-2'`):
  - `getTargetPath()`: used to populate the options in the `DocsVersionSwitcher` component
  - `getVersionFromPath()`: used to get the selected version to display as the `DocsVersionSwitcher` component 'label' (the version displayed to the user) and to remove it from the listed versions in the `DocsVersionSwitcher`  component so the user cannot select it again
  - `removeVersionFromPath()`: used to in the docs version alert banner (e.g `You are viewing documentation for version v202312-1. View latest version.`)
- Updated and added tests to ensure coverage of edge cases that were not originally included when docs version switcher was created

## 🤷 Why

In production, the following issues are happening on the TFE Releases docs:
1. On the TFE Release docs page, the dropdown is linking to invalid URLs, e.g. `/terraform/enterprise/v202312-1/releases/2024/v202401-1`
2. The dropdown was not updating to the show the selected version.
3. The dropdown doesn't show the latest version in the list, only on load.

## 🧪 Testing
###  🐛 🚫  Current behaviour in production 🚫 
1. **Dropdown is linking to the wrong URLs.**
- Visit this URL: https://developer.hashicorp.com/terraform/enterprise/releases/2024/v202401-1
- Now change the version to`v202311-1 `through the dropdown.
- It will 404 because the url that it is changing it to is: `v202311-1/releases/2023/v202311-1`, but it should be: `releases/2023/v202311-1`

2. **The dropdown is not updating to the show the current version.**
- Visit https://developer.hashicorp.com/terraform/enterprise/releases
- Select a version from the table (**do not** select a version from the dropdown), such as `v202312-1`
- See that the dropdown shows the latest version, not `v202312-1`

3. **The release pages don't show the "view latest release" banner**
Visit this page: https://developer.hashicorp.com/terraform/enterprise/releases/2023/v202311-1.
- Notice it doesn't have the version alert banner displayed

4. **The version alert banner does not link to the correct page**
- Visit this page: https://developer.hashicorp.com/terraform/enterprise/v202311-1/releases/2023/v202311-1.
- Click on "View latest version" link in the version alert banner. It will not take you to the latest releases page. Instead, it will take you to this page: https://developer.hashicorp.com/terraform/enterprise/releases/2023/v202311-1, which is the same URL except the first version (`v202311-1`) is removed from the URL.

---

### 🦋 🟢  How to test changes in this PR 🟢
1. Issues in production: **Dropdown is linking to the wrong URLs.** and **The dropdown is not updating to the show the current version.**
- Visit this URL: https://dev-portal-git-heat-fixtfe-dropdown-hashicorp.vercel.app/terraform/enterprise/releases).
- Now change the version to`v202311-1` through the dropdown.
- It should not return a 404 page, the path should be `/terraform/enterprise/releases/2023/v202311-1`, and the version in the dropdown should be `v202311-1`
- Hover over a link in the sidebar and there should be a version in the URL (see screenshot below)
<img width="559" alt="Screenshot 2024-02-12 at 18 25 56" src="https://github.com/hashicorp/dev-portal/assets/55773810/3c593419-815e-431b-b3a0-8c6e4291adee">

1. Issue in production:**The dropdown is not updating to the show the current version.**
- Visit https://dev-portal-git-heat-fixtfe-dropdown-hashicorp.vercel.app/terraform/enterprise/releases
- Select a version from the table (**do not** select a version from the dropdown), such as `v202312-1`
- See that the dropdown shows `v202312-1`, not the latest version (`v

1. Issue in production: **The release pages don't show the "view latest release" banner**
- Visit this page: https://dev-portal-git-heat-fixtfe-dropdown-hashicorp.vercel.app/terraform/enterprise/releases/2023/v202311-1.
- See that the version alert banner is displayed.
- Click on the "View latest version" link, which should take you to the [latest releases page](https://dev-portal-git-heat-fixtfe-dropdown-hashicorp.vercel.app/terraform/enterprise/releases) with the paths `/terraform/enterprise/releases`
